### ISSUE TITLE
Focus Trapping for Menu & Cart

### ISSUE THREAD

**Issue by gil--** [magento/pwa-studio#575]
## This issue is for the following packages:

[X] `venia-concept`
[ ] `pwa-buildpack`
[x] `peregrine`
[ ] `pwa-devdocs`
[ ] `upward-js`
[ ] `upward-spec`

## This issue is a:

[ ] Bug
[x] Feature suggestion
[ ] Documentation issue
[ ] Other (Please Specify)

## Description

When opening any of the overlay/modals such as cart, main menu, or filters, focus is not does not change to these "modals" and tabbing continues on main page behind these contents. This is not ideal for accessibility.

## Expected result:

When opening the mini cart, main menu, or other modal content, focus should be trapped, escape should close content, and tabbing should not continue outside of this content.

## Possible solutions:

1. We can use a 3rd party library like Reach-UI (dialog) which handles this for us: https://ui.reach.tech/dialog or implement our own focus handling.

2. Use [portals](https://reactjs.org/docs/portals.html#usage) and the [suggested ways](https://reactjs.org/docs/accessibility.html#programmatically-managing-focus) to handle focus.

### Here's an interactive demo of the current issue:

![focusbroken](https://user-images.githubusercontent.com/3484527/49299374-96a8a180-f48d-11e8-81c0-0a8089ef6a65.gif)

### Here is Reach UI where focus is trapped in the modal:

![focustrap](https://user-images.githubusercontent.com/3484527/49299447-c192f580-f48d-11e8-9202-f772f7815c76.gif)

**Comment by AlexeyKaryka** [comment]
seems like proposed https://ui.reach.tech/dialog is the most convenient way to handle focus within modals, there are number of css tricks and js libraries to trap focus within block, but I suppose library with react components is much more easier to integrate.

**Comment by brendanfalkowski** [comment]
I've done some work on focus trapping, and it can be a lose/lose situation depending on the context.

For example, you trigger a menu or modal and move focus to the first focus-able element in that container. But that element may not be the first element in the container — it's just the first element that can receive focus — so an assistive tech begins reading from the point of focus and ends up skipping non-focusable content within that container. 

Unless you can ensure the first focus-able element is also the first element in the trapped area, you've succeeded in hiding some content (usually a heading) from the user.

The `aria` spec for this suffers the same problem, and I haven't found any good solution so I've just avoided writing focus-trapping code unless it's a specifically built context. Hence lose/lose.

**Comment by sirugh** [comment]
Closing this as stale. We have done a lot of refactoring since this was opened and the menu and cart views have been reworked. We also have some open issues for handling these going forward.
